examples/nlp/nl_image_search.py

```diff
@@ -0,0 +1,635 @@
+
+feature_description = {
+    "caption": tf.io.FixedLenFeature([], tf.string),
+    "raw_image": tf.io.FixedLenFeature([], tf.string),
+}
+
+
+def read_example(example):
+    features = tf.io.parse_single_example(example, feature_description)
+    raw_image = features.pop("raw_image")
+    image = tf.image.decode_jpeg(raw_image, channels=3)
+    image = tf.image.resize(image, (299, 299))
+    features["image"] = image
+    return features
+
+
+def get_dataset(file_pattern, batch_size):
+
+    file_names = tf.data.Dataset.list_files(file_pattern)
+    dataset = tf.data.TFRecordDataset(file_names)
+    dataset = dataset.map(
+        read_example, num_parallel_calls=AUTOTUNE, deterministic=False
+    )
+    dataset = dataset.shuffle(batch_size * 10)
+    dataset = dataset.prefetch(buffer_size=AUTOTUNE)
+    dataset = dataset.batch(batch_size)
+
+    return dataset
+
+
+"""
+## Implement projection head
+
+The projection head is used to transform the image and the text embeddings to
+the same embedding space with the same dimensionality.
+"""
+
+
+def project_embeddings(
+    embeddings, num_projection_layers, projection_dims, dropout_rate
+):
+    projected_embeddings = layers.Dense(units=projection_dims)(embeddings)
+    for _ in range(num_projection_layers):
+        x = tf.nn.gelu(projected_embeddings)
+        x = layers.Dense(projection_dims)(x)
+        x = layers.Dropout(dropout_rate)(x)
+        x = layers.Add()([projected_embeddings, x])
+        projected_embeddings = layers.LayerNormalization()(x)
+    return projected_embeddings
+
+
+"""
+## Implement the vision encoder
+
+In this example, we use [Xception](https://keras.io/api/applications/xception/)
+from [Keras Applications](https://keras.io/api/applications/) as the base for the
+vision encoder.
+"""
+
+
+def create_vision_encoder(
+    num_projection_layers, projection_dims, dropout_rate, trainable=False
+):
+    # Load the pre-trained Xception model to be used as the base encoder.
+    xception = keras.applications.Xception(
+        include_top=False, weights="imagenet", pooling="avg"
+    )
+    # Set the trainability of the base encoder.
+    for layer in xception.layers:
+        layer.trainable = trainable
+    # Receive the images as inputs.
+    inputs = layers.Input(shape=(299, 299, 3), name="image_input")
+    # Preprocess the input image.
+    xception_input = tf.keras.applications.xception.preprocess_input(inputs)
+    # Generate the embeddings for the images using the xception model.
+    embeddings = xception(xception_input)
+    # Project the embeddings produced by the model.
+    outputs = project_embeddings(
+        embeddings, num_projection_layers, projection_dims, dropout_rate
+    )
+    # Create the vision encoder model.
+    encoder = keras.Model(inputs, outputs, name="vision_encoder")
+    return encoder
+
+
+"""
+## Implement the text encoder
+
+In this example, we use [BERT](https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-12_H-256_A-4/1)
```

Comment:
Suggestion:

Repetition

```suggestion
We use [BERT](https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-12_H-256_A-4/1)
```